spring-rabbit/src/main/java/org/springframework/amqp/rabbit/core/AsyncRabbitTemplate.java

```diff
@@ -0,0 +1,589 @@
+
+		private volatile ListenableFuture<Boolean> confirm;
+
+		private String nackCause;
+
+		public RabbitFuture(String correlationId) {
+			this.correlationId = correlationId;
+		}
+
+		@Override
+		public boolean cancel(boolean mayInterruptIfRunning) {
+			boolean cancelled = super.cancel(mayInterruptIfRunning);
+			pending.remove(this.correlationId);
+			return cancelled;
+		}
+
+		/**
+		 * When confirms are enabled contains a {@link ListenableFuture}
+		 * for the confirmation.
+		 * @return the future.
+		 */
+		public ListenableFuture<Boolean> getConfirm() {
+			return confirm;
+		}
+
+		void setConfirm(ListenableFuture<Boolean> confirm) {
+			this.confirm = confirm;
+		}
+
+		/**
+		 * When confirms are enabled and a nack is received, contains
+		 * the cause for the nack, if any.
+		 * @return the cause.
+		 */
+		public String getNackCause() {
+			return nackCause;
+		}
+
+		void setNackCause(String nackCause) {
+			this.nackCause = nackCause;
+		}
+
+	}
+
+	/**
+	 * A {@link RabbitFuture} with a return type of {@link Message}.
+	 *
+	 * @since 1.6
+	 *
+	 */
+	public class RabbitMessageFuture extends RabbitFuture<Message> {
+
+		public RabbitMessageFuture(String correlationId) {
+			super(correlationId);
+		}
+
+	}
+
+	/**
+	 * A {@link RabbitFuture} with a return type of the template's
+	 * generic parameter.
+	 *
+	 * @since 1.6
+	 *
+	 */
+	public class RabbitConverterFuture extends RabbitFuture<C> {
+
+		public RabbitConverterFuture(String correlationId) {
+			super(correlationId);
+		}
+
+	}
+
+	private final class CorrelationMessagePostProcessor implements MessagePostProcessor {
+
+		private final MessagePostProcessor userPostProcessor;
+
+		private final CorrelationData correlationData;
+
+		private volatile RabbitConverterFuture future;
+
+		public CorrelationMessagePostProcessor(MessagePostProcessor userPostProcessor,
+				CorrelationData correlationData) {
+			this.userPostProcessor = userPostProcessor;
+			this.correlationData = correlationData;
+		}
+
+		@Override
+		public Message postProcessMessage(Message message) throws AmqpException {
+			Message messageToSend = message;
+			if (this.userPostProcessor != null) {
+				messageToSend = this.userPostProcessor.postProcessMessage(message);
+			}
+			String correlationId = getOrSetCorrelationIdAndSetReplyTo(messageToSend);
+			this.future = new RabbitConverterFuture(correlationId);
+			if (this.correlationData instanceof MutableCorrelationData) {
+				((MutableCorrelationData) correlationData).setId(correlationId);
+				future.setConfirm(new SettableListenableFuture<Boolean>());
+			}
+			AsyncRabbitTemplate.this.pending.put(correlationId, future);
+			return messageToSend;
+		}
+
+		private RabbitConverterFuture getFuture() {
+			return this.future;
+		}
+
+	}
+
+	private static final class MutableCorrelationData extends CorrelationData {
+
+		public MutableCorrelationData(String id) {
+			super(id);
+		}
+
+		@Override
+		protected void setId(String id) {
```

Comment:
We need to document that we have this class to increase the `setId()` visibility here.
I was going to comment fully different... :flushed: